### PR TITLE
Refactor sstable part 2

### PR DESCRIPTION
**Pull request by thrawn01**
### Purpose
To move all the `sstable` marshaling code under the `internal/sstable` package.

### Implementation
- This PR builds upon the changes in #43
- Combined `DecodeBytesToSSTableInfo()` and `FlatBufferSSTableInfoCodec.Decode()` into `sstable.DecodeInfo()` which both decodes the flatbuf and preforms a checksum.
- `FlatBufferSSTableInfoCodec.Encode()` is now `sstable.EncodeInfo()` which avoids the copy from `sstable.Info` into `flatbuf.SsTableInfoT` and appends the checksum
- Removed `Info.SsTableIndex()` as it leaks flatbuffer implementation details outside of the `sstable` package by requiring users to call `.Unpack()` to get anything useful. Added `Info.BlockMeta()` and `Info.BlockMetaLength()` as a replacement.
- Changed signature of `sstable.NewBuilder()` to accept an `sstable.Config` which results in a constructor which looks like this
```go
builder := sstable.NewBuilder(sstable.Config{
    Compression:      compress.CodecNone,
    FilterBitsPerKey: 10,
    MinFilterKeys:    10,        
    BlockSize:        20,  
})
```
- Renamed `SsTableFormat` to `sstable.Config`
- Changed all methods on `SsTableFormat` to be package functions. All of the config needed to decode an SSTable should be contained in the SSTable it's self. The `sstable.Config` is only needed during encoding.

### Fixed leaky block abstraction
As a result of these changes, the `block` package is now responsible for encoding and decoding, compression and checksum of the entire block.

- Removed `SstInfoFromFlatBuf` as it was only used by `FlatBufferSSTableInfoCodec.Encode()`
- Discovered a decompression issue in `decodeBytesToBlock()`
- Removed `decodeBytesToBlock()` in favor of `block.Decode()` which now handles compression and checksum
- Modified `block.Encode()` to handle compression and checksum instead of `sstable.Builder` which should simplify how we use blocks.
- `FirstKey` now exists in `block.Block` instead of tracked by `sstable.Builder`.

### CODE REVIEW

File: internal/sstable/block/block.go [hunk 1]
```diff
@@ -45,10 +48,15 @@ type Block struct {
 // |  |  ...                                    |  |
 // |  +-----------------------------------------+  |
 // |  +-----------------------------------------+  |
+// |  |  Offsets for each KeyValue              |  |
```
Comment (naveen246) [inline internal/sstable/block/block.go:51]: This looks to be repeated, both `Block.Offsets` and `Offsets for each KeyValue` refer to the same I think

Comment (thrawn01) [reply internal/sstable/block/block.go]: thanks, fixed.

File: internal/sstable/flatbuf.go [hunk 2]
```diff
@@ -2,21 +2,37 @@ package sstable
...
+	info.sstableIndex = flatbuf.GetRootAsSsTableIndex(info.Data, 0)
+	info.blockMetaT = info.sstableIndex.UnPack().BlockMeta
+	return info.blockMetaT
 }
 
-func (info *Index) Size() int {
-	return len(info.Data)
+func (info *Index) BlockMetaLength() int {
+	if info.sstableIndex != nil {
+		info.sstableIndex.BlockMetaLength()
```
Comment (naveen246) [inline internal/sstable/flatbuf.go:32]: should be return info.sstableIndex.BlockMetaLength() ?

File: internal/sstable/decode.go [hunk 3]
```diff
@@ -138,21 +113,21 @@ func (f *SSTableFormat) getBlockRange(rng common.Range, sstInfo *Info, index *fl
...
-func (f *SSTableFormat) ReadBlocks(
+func ReadBlocks(
 	sstInfo *Info,
-	indexData *Index,
+	index *Index,
 	blockRange common.Range,
 	obj common.ReadOnlyBlob,
 ) ([]block.Block, error) {
-	index := indexData.SsTableIndex()
+	//index := indexData.SsTableIndex()
```
Comment (naveen246) [inline internal/sstable/decode.go:122]: this can be removed ?

File: internal/sstable/decode.go [hunk 4]
```diff
@@ -174,81 +149,43 @@ func (f *SSTableFormat) ReadBlocks(
...
-		f.CompressionCodec,
-	)
-}
-
-func (f *SSTableFormat) Clone() *SSTableFormat {
-	return &SSTableFormat{
-		BlockSize:        f.BlockSize,
-		MinFilterKeys:    f.MinFilterKeys,
-		CompressionCodec: f.CompressionCodec,
+	fmt.Printf("%+v\n", blockRange)
```
Comment (naveen246) [inline internal/sstable/decode.go:185]: should we use log.Print ?

**Pull request merged**